### PR TITLE
wait for analytics calls to finish in background

### PR DESCRIPTION
**Pull request by mdrohmann**
https://www.pivotaltracker.com/story/show/175624819

### CODE REVIEW

File: internal/analytics/analytics.go [hunk 1]
```diff
@@ -80,10 +81,18 @@ func (d *customDimensions) toMap() map[string]string {
...
+var (
+	eventWaitGroup sync.WaitGroup
+)
+
 func init() {
 	setup()
 }
 
+func WaitForAllEvents() {
+	eventWaitGroup.Wait()
```
Comment (Naatan) [inline internal/analytics/analytics.go:93]: Could we add a timeout to this? Let's wait for at most a second. I don't want slow connections or firewalls to cause commands to just hang. And loosing out on events from those outliers should be pretty negligible.

**Review by Naatan — CHANGES_REQUESTED** [review]

File: internal/analytics/analytics.go (hunk 1 above)
Comment (mdrohmann) [reply internal/analytics/analytics.go:109]: To add the timeout, I had to change our google-analytics library for this PR: 

So, we would need to merge this one first: https://github.com/ActiveState/go-ogle-analytics/pull/1, and then update the go.mod dependencies here.

File: internal/analytics/deferred.go [hunk 2]
```diff
@@ -21,6 +21,18 @@ type deferredData struct {
 
 const deferredCfgKey = "deferrer_analytics"
 
+func SetDeferred(da bool) {
```
Comment (Naatan) [inline internal/analytics/deferred.go:24]: @daved ran into this while addressing this story so added it here; `sendDeferred` was being called for commands where defer=true.

File: cmd/state/main.go [hunk 3]
```diff
@@ -82,6 +84,7 @@ func main() {
 		}
 	}
 
+	analytics.WaitForAllEvents(time.Second * 1)
```
Comment (daved) [inline cmd/state/main.go:87]: nit: I suggest putting this at the end of the captain command runner wrap that calls SetDefault (just to centralize the logic).

Comment (daved) [reply cmd/state/main.go:87]: Also nit: 2 or 3 more seconds should cover most circumstances where there is a long request hanging, but should not feel too bad to users (it should rarely happen).

File: internal/analytics/analytics.go [hunk 4]
```diff
@@ -82,11 +84,31 @@ func (d *customDimensions) toMap() map[string]string {
...
 	CustomDimensions = &customDimensions{}
 	setup()
 }
 
+// WaitForAllEvents waits for all events to return
+func WaitForAllEvents(t time.Duration) {
+	wg := make(chan bool)
+	go func() {
+		eventWaitGroup.Wait()
+		wg <- true
```
Comment (daved) [inline internal/analytics/analytics.go:101]: nit/not-nit:
```suggestion
		defer close(wg)
		eventWaitGroup.Wait()
```
This ensures resource cleanup.

File: internal/analytics/deferred.go [hunk 5]
```diff
@@ -21,6 +21,18 @@ type deferredData struct {
...
+	deferAnalytics = da
+	if deferAnalytics {
+		return
+	}
+	go func() {
+		if err := sendDeferred(sendEvent); err != nil {
+			logging.Errorf("Could not send deferred events: %v", err)
+		}
+	}()
```
Comment (daved) [inline internal/analytics/deferred.go:33]:
```suggestion
	eventWaitGroup.Add(1)
	go func() {
		defer eventWaitGroup.Done()
		if err := sendDeferred(sendEvent); err != nil {
			logging.Errorf("Could not send deferred events: %v", err)
		}
	}()
```
To be thorough (take as a nit if you'd like).

**Review by daved — CHANGES_REQUESTED** [review]

File: internal/analytics/analytics.go (hunk 4 above)
Comment (daved) [reply internal/analytics/analytics.go:101]: Also, it's idiomatic to use a `chan struct{}` since `struct{}` is a zero-sized primitive.

**Review by daved — APPROVED** [review]

**Pull request merged**